source/funkin/editors/character/CharacterEditor.hx

```diff
@@ -0,0 +1,700 @@
+	public var animAtlas:FlxAnimate;
+
+	public var cameraPointer:FlxSprite;
+
+	public var camFollow:FlxObject;
+	public var editorCamera:FlxCamera;
+	public var uiCamera:FlxCamera;
+
+	public var topMenu:Array<UIContextMenuOption>;
+	public var topMenuSpr:UITopMenu;
+
+	public var uiGroup:FlxTypedGroup<FlxSprite> = new FlxTypedGroup<FlxSprite>();
+	public var stage:FlxTypedSpriteGroup<FunkinSprite> = new FlxTypedSpriteGroup<FunkinSprite>(0, 0);
+
+	public var curAnim:Int = 0;
+
+	public var playingAnim:UIText;
+	public var movingCamText:UIText;
+
+	public var camZoom(default, set):Float = 1;
+
+	public function new(charName:String)
+	{
+		super();
+		this.curChar = charName;
+	}
+
+	override function create()
+	{
+		super.create();
+
+		// Camera setup
+		camFollow = new FlxObject(0, 0, 2, 2);
+		camFollow.screenCenter();
+		add(camFollow);
+		editorCamera = FlxG.camera;
+		uiCamera = new FlxCamera();
+		uiCamera.bgColor = 0;
+		FlxG.cameras.add(uiCamera);
+		FlxG.camera.follow(camFollow);
+		FlxCamera.defaultCameras = [editorCamera];
+
+		// Add Stage and Centered Cross
+		add(stage);
+		add(ofs_cross);
+		loadCross();
+
+		// Indicators
+		var pointer:FlxGraphic = FlxGraphic.fromClass(GraphicCursorCross);
+		cameraPointer = new FlxSprite().loadGraphic(pointer);
+		cameraPointer.setGraphicSize(40, 40);
+		cameraPointer.updateHitbox();
+		cameraPointer.color = FlxColor.WHITE;
+
+		// Character
+		add(charLayer);
+		add(cameraPointer);
+		loadChar(!CharacterConfig.loadCharacterData(curChar).isPlayer);
+
+		// Reload Stage
+		reloadStage();
+
+		// UI
+		loadUI();
+
+		playingAnim = new UIText((FlxG.width / 2), 30, -1, 'Animation:\n${char.characterData.animations[curAnim].name}', 24);
+		playingAnim.cameras = [uiCamera];
+
+		movingCamText = new UIText((FlxG.width / 4) - 310, (FlxG.height / 2), -1,
+			"Press the left mouse button anywhere on this stage to set the character's camera position.", 24);
+		movingCamText.cameras = [uiCamera];
+		movingCamText.visible = false;
+
+		// Top Menu
+		add(playingAnim);
+		add(movingCamText);
+		add(topMenuSpr);
+		add(uiGroup);
+	}
+
+	override function update(elapsed:Float)
+	{
+		// Control
+		controlCamera(elapsed);
+		controlCharacter();
+		changeAnim();
+		moveCharPos();
+		detectCamClick();
+
+		ghostChar.setPosition(char.x, char.y);
+		super.update(elapsed);
+	}
+
+	public function searchAnim(name:String, ?array:Array<AnimData>):AnimData
+	{
+		if (array == null)
+			array = char.characterData.animations;
+		for (i in 0...array.length - 1)
+			if (array[i].name == name)
+				return array[i];
+		return array[0];
+	}
+
+	public function getAnimIndex(name:String, ?array:Array<AnimData>):Int
+	{
+		if (array == null)
+			array = char.characterData.animations;
+		for (i in 0...array.length - 1)
+			if (array[i].name == name)
+				return i;
+		return 0;
+	}
+
+	function resortAnims(animations:Array<AnimData>)
+	{
+		if (getAnimIndex("idle") != 0)
+		{
+			var oldanims = animations;
```

Comment:
im pretty sure you need to call animations.copy()

Reply:
I mean that also works, so why change it?

Reply:
clear() is a inplace operation, meaning it causes issues, just do so to be safe